## CHICAGO CRIME DATA ANALYSIS
## Crime analysis of the city of Chicago from 2025 - 2026 based on the 250,000 most recent records

Data analysis pipeline
- 1. Data Loading from the API
- 2. Exploratory Data Analysis
- 3. Data Cleaning
- 4. SQL queries using SQL alchemy
- 5. Visualizations with Matplotlib and SEaborn
- 6. Mini Dashboard using tableau
- 7. Insights and recommendations

# 1. Imports and Setup

In [1]:
import matplotlib.pyplot as plt
from sqlalchemy import create_engine
import pandas as pd

# 2. Data Loading
- The data is sourced from the City of Chicago's website. It contains 250,000 reported crime incidents across 22 columns .

In [1]:
import pandas as pd
url = "https://data.cityofchicago.org/resource/ijzp-q8t2.csv?$limit=250000"
df = pd.read_csv(url)

In [2]:
df.to_csv('chicago_crime_data.csv', index = False)
df.head(2)

,id,case_number,date,block,iucr,primary_type,description,location_description,arrest,domestic,...,ward,community_area,fbi_code,x_coordinate,y_coordinate,year,updated_on,latitude,longitude,location
0,14183493,JK237735,2026-05-01T00:00:00.000,015XX N NORTH PARK AVE,0760,BURGLARY,BURGLARY FROM MOTOR VEHICLE,STREET,False,False,...,2.0,8.0,06,1173887.0,1910667.0,2026,2026-05-08T15:41:59.000,41.910283,-87.636638,"\n, \n(41.910282526, -87.636637837)"
1,14183496,JK237734,2026-05-01T00:00:00.000,014XX N NORTH PARK AVE,0760,BURGLARY,BURGLARY FROM MOTOR VEHICLE,STREET,False,False,...,2.0,8.0,06,1173910.0,1909937.0,2026,2026-05-08T15:41:59.000,41.908279,-87.636575,"\n, \n(41.908278858, -87.636575129)"


In [8]:
df = pd.read_csv('chicago_crime_data.csv')
df.head(5)

,id,case_number,date,block,iucr,primary_type,description,location_description,arrest,domestic,...,ward,community_area,fbi_code,x_coordinate,y_coordinate,year,updated_on,latitude,longitude,location
0,14183493,JK237735,2026-05-01T00:00:00.000,015XX N NORTH PARK AVE,0760,BURGLARY,BURGLARY FROM MOTOR VEHICLE,STREET,False,False,...,2.0,8.0,06,1173887.0,1910667.0,2026,2026-05-08T15:41:59.000,41.910283,-87.636638,"\n, \n(41.910282526, -87.636637837)"
1,14183496,JK237734,2026-05-01T00:00:00.000,014XX N NORTH PARK AVE,0760,BURGLARY,BURGLARY FROM MOTOR VEHICLE,STREET,False,False,...,2.0,8.0,06,1173910.0,1909937.0,2026,2026-05-08T15:41:59.000,41.908279,-87.636575,"\n, \n(41.908278858, -87.636575129)"
2,14187085,JK242258,2026-05-01T00:00:00.000,007XX W MONTROSE AVE,1130,DECEPTIVE PRACTICE,FRAUD OR CONFIDENCE GAME,NURSING / RETIREMENT HOME,False,False,...,46.0,3.0,11,1170527.0,1929439.0,2026,2026-05-08T15:41:59.000,41.961868,-87.648430,"\n, \n(41.961867922, -87.648430138)"
3,14184419,JK238980,2026-05-01T00:00:00.000,004XX W HURON ST,0910,MOTOR VEHICLE THEFT,AUTOMOBILE,STREET,False,False,...,42.0,8.0,07,1173197.0,1904992.0,2026,2026-05-08T15:41:59.000,41.894725,-87.639341,"\n, \n(41.894725375, -87.63934126)"
4,14189651,JK245347,2026-05-01T00:00:00.000,023XX N NATCHEZ AVE,1310,CRIMINAL DAMAGE,TO PROPERTY,RESIDENCE,False,False,...,36.0,19.0,14,NaN,NaN,2026,2026-05-08T15:41:59.000,NaN,NaN,NaN


In [5]:
print (df.shape)
print(df.head)
print(df.dtypes)

(250000, 22)
<bound method NDFrame.head of               id case_number                     date                   block  \
0       14183493    JK237735  2026-05-01T00:00:00.000  015XX N NORTH PARK AVE   
1       14183496    JK237734  2026-05-01T00:00:00.000  014XX N NORTH PARK AVE   
2       14187085    JK242258  2026-05-01T00:00:00.000    007XX W MONTROSE AVE   
3       14184419    JK238980  2026-05-01T00:00:00.000        004XX W HURON ST   
4       14189651    JK245347  2026-05-01T00:00:00.000     023XX N NATCHEZ AVE   
...          ...         ...                      ...                     ...   
249995  13797056    JJ207360  2025-04-05T14:30:00.000       048XX W NORTH AVE   
249996  13815942    JJ229877  2025-04-05T14:27:00.000       027XX W WILCOX ST   
249997  13797035    JJ207233  2025-04-05T14:20:00.000   051XX S LAKE PARK AVE   
249998  13796775    JJ207210  2025-04-05T14:20:00.000      070XX N RIDGE BLVD   
249999  13796811    JJ207266  2025-04-05T14:16:00.000         036X

# 3. Initial Exploratory Data Analysis
- Before cleaning we will examine the raw data to understand it. (Summary Statistics, missing values,distribution of crime types, location descriptions,)

In [9]:
#summary statistics 
df.describe()

,id,beat,district,ward,community_area,x_coordinate,y_coordinate,year,latitude,longitude
count,2.500000e+05,250000.000000,250000.000000,249999.000000,249991.000000,2.498750e+05,2.498750e+05,250000.000000,249875.000000,249875.000000
mean,1.396590e+07,1142.811188,11.197584,23.019488,36.238393,1.165773e+06,1.887202e+06,2025.282356,41.846054,-87.667156
std,6.086387e+05,708.089897,7.075346,13.949100,21.475234,1.622415e+04,3.131082e+04,0.450147,0.086103,0.059076
min,2.865500e+04,111.000000,1.000000,1.000000,1.000000,1.094182e+06,1.813909e+06,2025.000000,41.644608,-87.928903
25%,1.389518e+07,524.000000,5.000000,10.000000,22.000000,1.154568e+06,1.860302e+06,2025.000000,41.772153,-87.707930
50%,1.399097e+07,1032.000000,10.000000,23.000000,32.000000,1.167743e+06,1.893577e+06,2025.000000,41.863660,-87.659807
75%,1.408763e+07,1731.000000,17.000000,34.000000,53.000000,1.176929e+06,1.909960e+06,2026.000000,41.908518,-87.626224
max,1.419065e+07,2535.000000,31.000000,50.000000,77.000000,1.205119e+06,1.951495e+06,2026.000000,42.022548,-87.524529


In [ ]:
#null values
null_total = df.isnull().sum()
null_total

id                         0
case_number                0
date                       0
block                      0
iucr                       0
primary_type               0
description                0
location_description    1203
arrest                     0
domestic                   0
beat                       0
district                   0
ward                       1
community_area             9
fbi_code                   0
x_coordinate             125
y_coordinate             125
year                       0
updated_on                 0
latitude                 125
longitude                125
location                 125
dtype: int64

In [ ]:
#distribution of primary crimes
primary_crime_total = df['primary_type'].value_counts()
primary_crime_total

primary_type
THEFT                                57074
BATTERY                              45603
CRIMINAL DAMAGE                      28077
ASSAULT                              22912
MOTOR VEHICLE THEFT                  18855
OTHER OFFENSE                        17237
DECEPTIVE PRACTICE                   14811
BURGLARY                             11815
NARCOTICS                             7261
ROBBERY                               5733
CRIMINAL TRESPASS                     5729
WEAPONS VIOLATION                     5455
CRIMINAL SEXUAL ASSAULT               1754
OFFENSE INVOLVING CHILDREN            1639
SEX OFFENSE                           1388
PUBLIC PEACE VIOLATION                1151
INTERFERENCE WITH PUBLIC OFFICER       976
STALKING                               633
HOMICIDE                               462
ARSON                                  387
CONCEALED CARRY LICENSE VIOLATION      244
LIQUOR LAW VIOLATION                   205
PROSTITUTION                           19

In [14]:
#year distributions
year_dist = df['year'].value_counts().sort_index()
year_dist

year
2025    179411
2026     70589
Name: count, dtype: int64

In [19]:
arrest_rate = df['arrest'].mean()
print(f"{arrest_rate:.2%}")

15.37%


In [20]:
domestic_crime_rate = df['domestic'].mean()
print(f"{domestic_crime_rate:.2%}")

19.14%


In [22]:
top_locations = df['location_description'].value_counts().head(10)
top_locations

location_description
STREET                                    67243
APARTMENT                                 48779
RESIDENCE                                 29176
SIDEWALK                                  12579
PARKING LOT / GARAGE (NON RESIDENTIAL)     9181
SMALL RETAIL STORE                         9091
DEPARTMENT STORE                           5917
RESTAURANT                                 5562
ALLEY                                      5358
OTHER (SPECIFY)                            4467
Name: count, dtype: int64

# AND .Data Cleaning
- Dealing with missing values;
1. Date parsing
2. Duplicate rows
3. Text standardization
4. Location description column
5. Handling null values
6. Feature engineering

In [27]:
# 1. Date parsing (from string to datetime)
df['date'] = pd.to_datetime(df['date'])
df['date'].dtype
df['hour'] = df['date'].dt.hour
df['day_name'] = df['date'].dt.day_name()
print(df['date'].dtype)

df['date'].head(3).tolist()

datetime64[ns]


[Timestamp('2026-05-01 00:00:00'),
 Timestamp('2026-05-01 00:00:00'),
 Timestamp('2026-05-01 00:00:00')]

In [29]:
# 2. Duplicate records
df.duplicated().sum()

0

In [30]:
# 3. Text standardization
text_cols = ['primary_type', 'description', 'location_description', 'block', 'fbi_code']

for col in text_cols:
    before_unique = df[col].nunique()
    df[col] = df[col].str.strip().str.upper()
    after_unique = df[col].nunique()
    print(f'{col}: unique values {before_unique} -> {after_unique}')
    

primary_type: unique values 31 -> 31
description: unique values 312 -> 312
location_description: unique values 130 -> 130
block: unique values 28352 -> 26910
fbi_code: unique values 26 -> 26


In [32]:
# 4. Location description column
consolidations = {
    'OTHER (SPECIFY)': 'OTHER',
    'OTHER COMMERCIAL TRANSPORTATION': 'OTHER TRANSPORTATION',
}

before_unique = df['location_description'].nunique()
df['location_description'] = df['location_description'].replace(consolidations)
after_unique = df['location_description'].nunique()

print(f'location_description unique values: {before_unique} -> {after_unique}')
print(f'Rows affected by consolidation: {df["location_description"].isin(consolidations.values()).sum():,}')


location_description unique values: 130 -> 130
Rows affected by consolidation: 4,553


In [33]:
# null values handling
df['location_description'] = df['location_description'].fillna('UNKNOWN')
df['ward'] = df['ward'].fillna(df['ward'].median())
df['community_area'] = df['community_area'].fillna(df['community_area'].median())

In [34]:
df.isnull().sum()

id                        0
case_number               0
date                      0
block                     0
iucr                      0
primary_type              0
description               0
location_description      0
arrest                    0
domestic                  0
beat                      0
district                  0
ward                      0
community_area            0
fbi_code                  0
x_coordinate            125
y_coordinate            125
year                      0
updated_on                0
latitude                125
longitude               125
location                125
hour                      0
day_name                  0
dtype: int64

In [39]:
# 6. Feature engineering with date time
day_order = ['Monday', 'Tuesday', 'Wednesday', 'Thursday', 'Friday', 'Saturday', 'Sunday']

df['month'] = df['date'].dt.month
df['hour'] = df['date'].dt.hour
df['day_of_week'] = pd.Categorical(
    df['date'].dt.day_name(),
    categories=day_order,
    ordered=True
)
df['quarter'] = df['date'].dt.quarter
df['midnight_flag'] = df['hour'] == 0

df['date'].dtype

dtype('<M8[ns]')

In [40]:
df.to_csv('chicago_data_cleaned.csv', index=False)

## 4. DATA ANALYSIS WITH SQL

# We'll load the cleaned DataFrame into sqlalchemy

In [3]:
df2 = pd.read_csv('chicago_data_cleaned.csv')
df2.head(2)

,id,case_number,date,block,iucr,primary_type,description,location_description,arrest,domestic,...,updated_on,latitude,longitude,location,hour,day_name,month,day_of_week,quarter,midnight_flag
0,14183493,JK237735,2026-05-01 00:00:00,015XX N NORTH PARK AVE,0760,BURGLARY,BURGLARY FROM MOTOR VEHICLE,STREET,False,False,...,2026-05-08T15:41:59.000,41.910283,-87.636638,"\r\n, \r\n(41.910282526, -87.636637837)",0,Friday,5,Friday,2,True
1,14183496,JK237734,2026-05-01 00:00:00,014XX N NORTH PARK AVE,0760,BURGLARY,BURGLARY FROM MOTOR VEHICLE,STREET,False,False,...,2026-05-08T15:41:59.000,41.908279,-87.636575,"\r\n, \r\n(41.908278858, -87.636575129)",0,Friday,5,Friday,2,True


In [18]:
engine = create_engine('sqlite:///chicago_crime.db')
df2.to_sql('chicago_crimes', engine, index=False, if_exists='replace')


250000

In [19]:
count = pd.read_sql('SELECT COUNT(*) as total FROM chicago_crimes', engine)['total']
count

0    250000
Name: total, dtype: int64

# Queries
- Top 10 crimes by volume with percentage
- Arrest rate by crime type
- Domestic vs non-domestic comparisons
- Crime volume vs arrest rate by police district
- Monthly crime counts
- Top 10 locations type by incident count

In [21]:
# Query 1: Top 10 crime types by volume with percentage share
q1 = pd.read_sql("""
    SELECT
        primary_type,
        COUNT(*) AS total_incidents,
        ROUND(100.0 * COUNT(*) / (SELECT COUNT(*) FROM chicago_crimes), 2) AS pct_of_total
    FROM chicago_crimes
    GROUP BY primary_type
    ORDER BY total_incidents DESC
    LIMIT 10
""", engine)
print('Top 10 crime types:')
q1

Top 10 crime types:


,primary_type,total_incidents,pct_of_total
0,THEFT,57074,22.83
1,BATTERY,45603,18.24
2,CRIMINAL DAMAGE,28077,11.23
3,ASSAULT,22912,9.16
4,MOTOR VEHICLE THEFT,18855,7.54
5,OTHER OFFENSE,17237,6.89
6,DECEPTIVE PRACTICE,14811,5.92
7,BURGLARY,11815,4.73
8,NARCOTICS,7261,2.90
9,ROBBERY,5733,2.29


- Insights : The most committed crime is theft with a total of 57,074 incidents which is making 22.83% of all crimes commited in Chicago

In [22]:
# Query 2: Arrest rate by crime type (types with more than 1,000 incidents)
q2 = pd.read_sql("""
    SELECT
        primary_type,
        COUNT(*) AS total_incidents,
        SUM(CASE WHEN arrest = 1 THEN 1 ELSE 0 END) AS total_arrests,
        ROUND(
            100.0 * SUM(CASE WHEN arrest = 1 THEN 1 ELSE 0 END) / COUNT(*),
            2
        ) AS arrest_rate_pct
    FROM chicago_crimes
    GROUP BY primary_type
    HAVING total_incidents > 1000
    ORDER BY arrest_rate_pct DESC
    LIMIT 10
""", engine)
print('Arrest rate by crime type:')
q2

Arrest rate by crime type:


,primary_type,total_incidents,total_arrests,arrest_rate_pct
0,NARCOTICS,7261,6833,94.11
1,WEAPONS VIOLATION,5455,4203,77.05
2,PUBLIC PEACE VIOLATION,1151,591,51.35
3,CRIMINAL TRESPASS,5729,1843,32.17
4,BATTERY,45603,8619,18.90
5,OTHER OFFENSE,17237,3126,18.14
6,ASSAULT,22912,2871,12.53
7,ROBBERY,5733,537,9.37
8,THEFT,57074,5043,8.84
9,SEX OFFENSE,1388,105,7.56


Insights : Narcotics has the highest arrest percentage, even though it has one of the lowest total incidents
- Sex offence crimes has the lowest arrest percentage with a total of 105 arrests out of 1388 total incidents.
- Theft has the highest total incidents yet a small arrest percentage of 8.84%

In [23]:
# Query 3: Domestic vs non-domestic crime breakdown
q3 = pd.read_sql("""
    SELECT
        CASE WHEN domestic = 1 THEN 'Domestic' ELSE 'Non-Domestic' END AS incident_type,
        COUNT(*) AS total,
        ROUND(100.0 * COUNT(*) / (SELECT COUNT(*) FROM chicago_crimes), 2) AS pct
    FROM chicago_crimes
    GROUP BY domestic
""", engine)
print('Domestic vs Non-Domestic:')
q3

Domestic vs Non-Domestic:


,incident_type,total,pct
0,Non-Domestic,202143,80.86
1,Domestic,47857,19.14


Non- Domestic incidents make up most of the incident type with 80.86% of all incidents

In [24]:
# Query 4: Crime volume and arrest rate by police district
q4 = pd.read_sql("""
    SELECT
        district,
        COUNT(*) AS total_crimes,
        SUM(CASE WHEN arrest = 1 THEN 1 ELSE 0 END) AS total_arrests,
        ROUND(
            100.0 * SUM(CASE WHEN arrest = 1 THEN 1 ELSE 0 END) / COUNT(*),
            2
        ) AS arrest_rate_pct
    FROM chicago_crimes
    GROUP BY district
    ORDER BY total_crimes DESC
    LIMIT 10
""", engine)
print('Top districts by crime volume:')
q4

Top districts by crime volume:


,district,total_crimes,total_arrests,arrest_rate_pct
0,12,16318,1585,9.71
1,8,15736,2295,14.58
2,1,14378,2517,17.51
3,6,14204,2005,14.12
4,18,13929,1943,13.95
5,3,13790,1961,14.22
6,2,13771,1499,10.89
7,4,13752,1737,12.63
8,19,13448,1206,8.97
9,11,13262,3620,27.30


Insights : The 12th  District has the most total crimes but with a low arrest rate as compared to the 11th district which has the lowest total crimes but with the highest arrest rate as compared to other districts

In [25]:
# Query 5: Monthly crime counts
q5 = pd.read_sql("""
    SELECT
        month,
        COUNT(*) AS incidents
    FROM chicago_crimes
    GROUP BY month
    ORDER BY incidents desc
""", engine)
print('Monthly crime counts:')
q5

Monthly crime counts:


,month,incidents
0,4,35189
1,7,22651
2,8,21323
3,6,21108
4,10,20962
5,5,20502
6,9,20330
7,3,18860
8,11,18392
9,12,17462


Insights : April has the most total incidents and February has the lowest total incidents

In [26]:
# Query 6: Top 10 location types by incident count
q6 = pd.read_sql("""
    SELECT
        location_description,
        COUNT(*) AS incidents
    FROM chicago_crimes
    GROUP BY location_description
    ORDER BY incidents DESC
    LIMIT 10
""", engine)
print('Top location types:')
q6

Top location types:


,location_description,incidents
0,STREET,67243
1,APARTMENT,48779
2,RESIDENCE,29176
3,SIDEWALK,12579
4,PARKING LOT / GARAGE (NON RESIDENTIAL),9181
5,SMALL RETAIL STORE,9091
6,DEPARTMENT STORE,5917
7,RESTAURANT,5562
8,ALLEY,5358
9,OTHER,4467


Insights : Most crime incidents happen in the street then followed by apartments